app/src/Controllers/ZipController.php

```diff
@@ -33,41 +36,76 @@ public function __invoke(Request $request, Response $response): ResponseInterfac
             return $response->withStatus(404, $this->translator->trans('error.file_not_found'));
         }
 
-        $response->getBody()->write(
-            $this->cache->get(sprintf('zip-%s', sha1($path)), function () use ($path): string {
-                return $this->createZip($path)->getContents();
-            })
-        );
-
-        return $response->withHeader('Content-Type', 'application/zip')
+        $response = $response
+            ->withHeader('Content-Type', 'application/zip')
             ->withHeader('Content-Disposition', sprintf(
                 'attachment; filename="%s.zip"',
                 $this->generateFileName($path)
-            ));
+            ))
+            ->withHeader('X-Accel-Buffering', 'no');
+
+        $files = $this->finder->in($path)->files();
+
+        $zip = $this->createZip($path, $files);
+        $size = $zip->finish();
+
+        $response = $this
+        ->augmentHeadersWithEstimatedSize($response, $size)
+        ->withBody(new CallbackStream(function () use ($zip) {
+            $zip->executeSimulation();
+        }));
+
+        return $response;
     }
 
-    /** Create a zip file from a directory. */
-    protected function createZip(string $path): TemporaryFile
+    /** Create a zip stream from a directory.
+     *
+     * @throws \ZipStream\Exception\FileNotFoundException
+     * @throws \ZipStream\Exception\FileNotReadableException
+     */
+    protected function createZip(string $path, Finder $files): ZipStream
     {
-        $zip = new ZipArchive;
-        $zip->open((string) $tempFile = new TemporaryFile(
-            $this->config->get('cache_path')
-        ), ZipArchive::CREATE | ZipArchive::OVERWRITE);
+        $compressionMethod = $this->config->get('zip_compress') ? CompressionMethod::DEFLATE : CompressionMethod::STORE;
+
+        $zip = new ZipStream(
+            sendHttpHeaders: false,
+            operationMode: OperationMode::SIMULATE_LAX
+        );
 
-        foreach ($this->finder->in($path)->files() as $file) {
-            $zip->addFile((string) $file->getRealPath(), $this->stripPath($file, $path));
+        foreach ($files as $file) {
+            $modifiedTime = null;
+
+            try {
+                $modifiedTime = new \DateTime('@' . (int) $file->getMTime());
+            } catch (\Exception $e) {
+                // We couldn't get the creation time, so we don't set it
+            }
```

Comment:
We can use the same approach used in `ModifiedTime.php` here and use `lstat()` as a fall back.

https://github.com/DirectoryLister/DirectoryLister/blob/10795b101167044331e66ddae739d640f99c4a7d/app/src/ViewFunctions/ModifiedTime.php#L20-L24

```suggestion
            try {
                $modifiedTime = new \DateTime('@' . (int) $file->getMTime());
            } catch (RuntimeException $e) {
                $modifiedTime = lstat($file->getPathname())['mtime'];
            }
```